# nf-core/grzqc: Pipeline Overview and Output

## Introduction

This document describes some key steps as well as the output produced by the pipeline. 

## Pipeline overview

The pipeline is built using [Nextflow](https://www.nextflow.io/) and processes data using the following steps:

- Read QC ([`FastQC`](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/) and [`FASTP`](https://github.com/OpenGene/fastp))
- Alignment using ([`BWAMEM2`](https://github.com/bwa-mem2/bwa-mem2))
- MarkDuplicates using ([`Picard`](https://gatk.broadinstitute.org/hc/en-us/articles/360037052812-MarkDuplicates-Picard))
- Coverage calculation by ([`Mosdepth`](https://github.com/brentp/mosdepth))
- Present QC for raw reads ([`MultiQC`](http://multiqc.info/))
- Report metrics generated during the workflow execution [`Pipeline information`](#pipeline-information) 

For the exact command lines executed by the pipeline you may check the workflow reports automatically generated by the test pipeline on GitHub.
To access these reports, click on the title of the latest successful [pipeline run](https://github.com/BfArM-MVH/GRZ_QC_Workflow/actions/workflows/main.yml) and download the `nextflow-pipeline-info` artifact under the "Artifacts" section at the bottom of the page.
The command lines are detailed the the "Tasks" table at the bottom of the `execution_report_*.html` file.

## Pipeline output

### mosdepth

[Mosdepth](https://github.com/brentp/mosdepth) reports information for the evaluation of the quality of the provided alignment data.
In short, the basic statistics of the alignment (number of reads, coverage, GC-content, etc.) are summarized and a number of useful graphs are produced.

The pipeline uses branching logic to provide the correct BED file depending on the library type:

- For WGS samples:
The pipeline automatically selects the default BED file containing ~400 genes for the mosdepth run. Mosdepth calculates depths for both genome-wide and targeted regions in one run.

- For WES/Panel samples:
The pipeline uses the target region BED file supplied in the submission metadata and converts it into UCSC-style if the chromosome names are in NCBI-style.

| **Library Type** | **input: BED File**                   | **Mean Coverage Calculation (`meanDepthOfCoverage`)**           | **per-base Coverage of Targeted Regions (to furthur calculate `fractionBasesAboveQualityThresholdRequired`)**                  |
|------------------|------------------------------------|-----------------------------------------|----------------------------------------|
| **WGS**          | Default representative genes, ([hg19](../assets/default_files/hg19_439_omim_genes.bed), [hg38](../assets/default_files/hg38_440_omim_genes.bed)) |  genome-wide average coverage           | Computed over a predefined set of ~400 gene regions         |
| **WES/Panel**    | User-provided BED file                         | Average coverage on user-specified targets | Computed over the user-provided target regions              |

<details markdown="1">
<summary>Output files for all samples</summary>

**Output directory: `{outdir}/results/mosdepth/<sample>`**

- `<sample>.mosdepth.summary.txt`
  - A summary of mean depths per chromosome and within specified regions per chromosome.  
- `<sample>.regions.bed.gz`
  - per-base depth for chromosome and genome-wide, or targeted regions

  </details>

### MultiQC

[MultiQC](http://multiqc.info) is a visualization tool that generates a single HTML report summarising all samples in your project. Most of the pipeline QC results are visualised in the report and further statistics are available in the report data directory.

Results generated by MultiQC collate pipeline QC from supported tools e.g. FastQC. The pipeline has special steps which also allow the software versions to be reported in the MultiQC output for future traceability. For more information about how to use MultiQC reports, see <http://multiqc.info>.

<details markdown="1">
<summary>Output files</summary>

- `multiqc/`
  - `multiqc_report.html`: a standalone HTML file that can be viewed in your web browser.
  - `multiqc_data/`: directory containing parsed statistics from the different tools used in the pipeline.
  - `multiqc_plots/`: directory containing static images from the report in various formats.

</details>

### Pipeline information

<details markdown="1">
<summary>Output files</summary>

- `pipeline_info/`
  - Reports generated by Nextflow: `execution_report.html`, `execution_timeline.html`, `execution_trace.txt` and `pipeline_dag.dot`/`pipeline_dag.svg`.
  - Reports generated by the pipeline: `pipeline_report.html`, `pipeline_report.txt` and `software_versions.yml`. The `pipeline_report*` files will only be present if the `--email` / `--email_on_fail` parameter's are used when running the pipeline.
  - Reformatted samplesheet files used as input to the pipeline: `samplesheet.valid.csv`.
  - Parameters used by the pipeline run: `params.json`.

</details>

[Nextflow](https://www.nextflow.io/docs/latest/tracing.html) provides excellent functionality for generating various reports relevant to the running and execution of the pipeline. This will allow you to troubleshoot errors with the running of the pipeline, and also provide you with other information such as launch commands, run times and resource usage.
